bokehjs/src/lib/core/selection_manager.ts

```diff
@@ -1,35 +1,21 @@
-import {HasProps} from "./has_props"
 import {Geometry} from "./geometry"
 import {SelectionMode} from "core/enums"
 import {Selection} from "models/selections/selection"
 import type {ColumnarDataSource} from "models/sources/columnar_data_source"
 import {DataRenderer, DataRendererView} from "models/renderers/data_renderer"
-import {GlyphRendererView} from "models/renderers/glyph_renderer"
-import {GraphRendererView} from "models/renderers/graph_renderer"
-import * as p from "./properties"
+import type {GlyphRendererView} from "models/renderers/glyph_renderer"
+import type {GraphRendererView} from "models/renderers/graph_renderer"
 
-export namespace SelectionManager {
-  export type Props = HasProps.Props & {
-    source: p.Property<ColumnarDataSource>
-  }
-
-  export type Attrs = p.AttrsOf<Props>
+// XXX: this is needed to cut circular dependency between this, models/renderers/* and models/sources/*
+function is_GlyphRendererView(renderer_view: DataRendererView): renderer_view is GlyphRendererView {
+  return renderer_view.model.type == "GlyphRenderer"
+}
+function is_GraphRendererView(renderer_view: DataRendererView): renderer_view is GraphRendererView {
+  return renderer_view.model.type == "GraphRenderer"
```

Comment:
Changes like this are unfortunate and it would have been much better to just keep `instanceof` checks, but this is the only way right now to cut those cycles. Ultimately selection management has to be redesigned, so that selection manager doesn't know about any specific types of data renderers, which will ultimately cut the cycle by design. The current design doesn't make sense from (at least) extensibility perspective, because adding a new type of data renderer requires changing `SelectionManager`, among other parts of bokehjs.

Reply:
Agree that the selection manager ought to delegate more operations to avoid this kind of coupling.